lib/msal-node/src/client/ClientCredentialClient.ts

```
@@ -89,7 +89,26 @@ export class ClientCredentialClient extends BaseClient {
             this.serverTelemetryManager?.setCacheOutcome(
                 CacheOutcome.CACHED_ACCESS_TOKEN_EXPIRED
             );
+
             return null;
+        // the token is not expired, but it must be refreshed due to the refresh_in value
+        } else if (
```

The approach is good, but the structure of the code is not. These private methods should do one thing and one thing alone, .e.g. `getCachedAuthenticationResult` should touch the cache, but now it also makes HTTP requests.

A possible implementation:

- `acquireToken` is the orchestrator of the logic
- it gets a cached access token just like before and checks the CacheOutcome
- if the CacheOutcome is REFRESH_CACHED_ACCESS_TOKEN then it keep hold of the "oldAuthenticationResult" and it calls `executeTokenRequest`
- if `executeTokenRequest` succeeds, return its result
- if `executeTokenRequest` fails for any reason other than 4xx errors, it returns `oldAuthenticationResult`
- if `executeTokenRequest` fails for 4xx errors, let the exception bubble through